File: lib/data/models/transaction.dart
import 'tracking_log.dart';

class Transaction {
  final String id;
  final String noResi;
  final String kodeGerai;
  final String barcodeData;
  final Map<String, dynamic> pengirim;
  final Map<String, dynamic> penerima;
  final Map<String, dynamic> paket;
  final Map<String, dynamic> createdBy;
  final String statusSaatIni;
  final String? namaDriver;
  final String? kontakDriver;
  final String? driverUserId;
  final String? noManifest;
  final Map<String, dynamic>? tujuanSelanjutnya;
  final String? namaPenerimaAkhir;
  final String? currentCabangId;
  final Map<String, dynamic>? lokasiPenerima;
  final String? jenisMasalah;
  final String? catatanMasalah;
  final Map<String, dynamic>? dilaporkanOleh;
  final DateTime? dilaporkanPada;
  final Map<String, dynamic>? diselesaikanOleh;
  final DateTime? diselesaikanPada;
  final String jenisPembayaran;
  final String statusPembayaran;
  final int tempoHari;
  final Map<String, dynamic>? pembayaranDikonfirmasiOleh;
  final DateTime? pembayaranDikonfirmasiPada;
  final String? codCabangId;
  final List<TrackingLog> trackingLogs;
  final DateTime createdAt;
  final DateTime updatedAt;

  Transaction({
    required this.id,
    required this.noResi,
    required this.kodeGerai,
    required this.barcodeData,
    required this.pengirim,
    required this.penerima,
    required this.paket,
    required this.createdBy,
    required this.statusSaatIni,
    this.namaDriver,
    this.kontakDriver,
    this.driverUserId,
    this.noManifest,
    this.tujuanSelanjutnya,
    this.namaPenerimaAkhir,
    this.currentCabangId,
    this.lokasiPenerima,
    this.jenisMasalah,
    this.catatanMasalah,
    this.dilaporkanOleh,
    this.dilaporkanPada,
    this.diselesaikanOleh,
    this.diselesaikanPada,
    this.jenisPembayaran = 'cash',
    this.statusPembayaran = 'paid',
    this.tempoHari = 14,
    this.pembayaranDikonfirmasiOleh,
    this.pembayaranDikonfirmasiPada,
    this.codCabangId,
    required this.trackingLogs,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Transaction.fromJson(Map<String, dynamic> json) => Transaction(
    id: json['_id'] as String,
    noResi: json['no_resi'] as String,
    kodeGerai: json['kode_gerai'] as String,
    barcodeData: json['barcode_data'] as String,
    pengirim: Map<String, dynamic>.from(json['pengirim'] as Map),
    penerima: Map<String, dynamic>.from(json['penerima'] as Map),
    paket: Map<String, dynamic>.from(json['paket'] as Map),
    createdBy: Map<String, dynamic>.from(json['created_by'] as Map),
    statusSaatIni: json['status_saat_ini'] as String,
    namaDriver: json['nama_driver'] as String?,
    kontakDriver: json['kontak_driver'] as String?,
    driverUserId: json['driver_user_id'] as String?,
    noManifest: json['no_manifest'] as String?,
    tujuanSelanjutnya: json['tujuan_selanjutnya'] as Map<String, dynamic>?,
    namaPenerimaAkhir: json['nama_penerima_akhir'] as String?,
    currentCabangId: json['current_cabang_id'] as String?,
    lokasiPenerima: json['lokasi_penerima'] as Map<String, dynamic>?,
    jenisMasalah: json['jenis_masalah'] as String?,
    catatanMasalah: json['catatan_masalah'] as String?,
    dilaporkanOleh: json['dilaporkan_oleh'] as Map<String, dynamic>?,
    dilaporkanPada: json['dilaporkan_pada'] != null ? DateTime.parse(json['dilaporkan_pada'] as String) : null,
    diselesaikanOleh: json['diselesaikan_oleh'] as Map<String, dynamic>?,
    diselesaikanPada: json['diselesaikan_pada'] != null ? DateTime.parse(json['diselesaikan_pada'] as String) : null,
    jenisPembayaran: json['jenis_pembayaran'] as String? ?? 'cash',
    statusPembayaran: json['status_pembayaran'] as String? ?? 'paid',
    tempoHari: json['tempo_hari'] as int? ?? 14,
    pembayaranDikonfirmasiOleh: json['pembayaran_dikonfirmasi_oleh'] as Map<String, dynamic>?,
    pembayaranDikonfirmasiPada: json['pembayaran_dikonfirmasi_pada'] != null ? DateTime.parse(json['pembayaran_dikonfirmasi_pada'] as String) : null,
    codCabangId: json['cod_cabang_id'] as String?,
    trackingLogs: (json['tracking_logs'] as List<dynamic>?)
        ?.map((e) => TrackingLog.fromJson(Map<String, dynamic>.from(e as Map)))
        .toList() ?? [],
    createdAt: DateTime.parse(json['createdAt'] as String),
    updatedAt: DateTime.parse(json['updatedAt'] as String),
  );

  String get pengirimName => pengirim['name'] as String? ?? '';
  String get penerimaName => penerima['name'] as String? ?? '';
  String get penerimaAddress => penerima['address'] as String? ?? '';
  double? get penerimaLatitude {
    final coords = lokasiPenerima?['coordinates'] as List<dynamic>?;
    if (coords != null && coords.length >= 2) return (coords[1] as num).toDouble();
    return null;
  }
  double? get penerimaLongitude {
    final coords = lokasiPenerima?['coordinates'] as List<dynamic>?;
    if (coords != null && coords.length >= 2) return (coords[0] as num).toDouble();
    return null;
  }
  String get beratLabel => '${paket['berat_kg']?.toStringAsFixed(1) ?? '0'} kg';
  String get koliLabel => '${paket['jumlah_koli'] ?? '0'} koli';
  int get jumlahKoli => (paket['jumlah_koli'] as num?)?.toInt() ?? 0;
  double get biayaKirim => (paket['biaya_kirim'] as num?)?.toDouble() ?? 0;

  /// Nama cabang tempat paket diterima (dari tracking log terakhir status diterima_cabang)
  String get diterimaDiCabang {
    if (statusSaatIni != 'diterima_cabang') return '';
    for (final log in trackingLogs.reversed) {
      if (log.status == 'diterima_cabang') {
        final nama = log.lokasiName;
        if (nama.isNotEmpty) return nama;
        return log.tujuan?['nama'] as String? ?? '';
      }
    }
    return '';
  }
}

File: backend/models/Transaction.js
const mongoose = require('mongoose');

const trackingLogSchema = new mongoose.Schema({
  status: {
    type: String,
    enum: ['diterima_cabang', 'keluar_cabang', 'proses_kirim', 'diterima', 'hilang', 'gagal_kirim', 'kasus_selesai'],
    required: true,
  },
  deskripsi: { type: String, default: '' },
  pelaku: {
    user_id: { type: mongoose.Schema.Types.ObjectId, ref: 'User', default: null },
    name: { type: String, required: true },
    role: { type: String, required: true },
  },
  lokasi: {
    nama: { type: String, default: '' },
    tipe: { type: String, enum: ['cabang', ''], default: '' },
    cabang_id: { type: mongoose.Schema.Types.ObjectId, ref: 'Cabang', default: null },
  },
  driver_ditugaskan: {
    user_id: { type: mongoose.Schema.Types.ObjectId, ref: 'User', default: null },
    nama: { type: String, default: '' },
    kontak: { type: String, default: '' },
  },
  tujuan: {
    tipe: { type: String, enum: ['cabang', 'penerima', ''], default: '' },
    nama: { type: String, default: '' },
  },
  nama_penerima: { type: String, default: '' },
  no_manifest: { type: String, default: '' },
  timestamp: { type: Date, default: Date.now },
});

const transactionSchema = new mongoose.Schema({
  no_resi: { type: String, required: true, unique: true },
  kode_gerai: { type: String, required: true },
  barcode_data: { type: String, required: true },

  pengirim: {
    name: { type: String, required: true },
    phone: { type: String, required: true },
    address: { type: String, required: true },
  },
  penerima: {
    name: { type: String, required: true },
    phone: { type: String, required: true },
    address: { type: String, required: true },
    kecamatan: { type: String, default: '' },
    kota: { type: String, default: '' },
  },
  lokasi_penerima: {
    type: { type: String, enum: ['Point'], default: 'Point' },
    coordinates: { type: [Number], default: null },
  },
  paket: {
    berat_kg: { type: Number, required: true },
    jumlah_koli: { type: Number, required: true },
    biaya_kirim: { type: Number, required: true },
  },

  created_by: {
    user_id: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
    name: { type: String, required: true },
    role: { type: String, required: true },
    cabang_id: { type: mongoose.Schema.Types.ObjectId, ref: 'Cabang', default: null },
    cabang_name: { type: String, default: '' },
  },

  status_saat_ini: {
    type: String,
    enum: ['diterima_cabang', 'keluar_cabang', 'proses_kirim', 'diterima', 'hilang', 'gagal_kirim', 'kasus_selesai'],
    default: 'diterima_cabang',
  },

  jenis_masalah: { type: String, enum: ['hilang', 'gagal_kirim', null], default: null },
  catatan_masalah: { type: String, default: '' },
  dilaporkan_oleh: {
    user_id: { type: mongoose.Schema.Types.ObjectId, ref: 'User', default: null },
    name: { type: String, default: '' },
    role: { type: String, default: '' },
    cabang_id: { type: mongoose.Schema.Types.ObjectId, ref: 'Cabang', default: null },
    cabang_name: { type: String, default: '' },
  },
  dilaporkan_pada: { type: Date, default: null },
  diselesaikan_oleh: {
    user_id: { type: mongoose.Schema.Types.ObjectId, ref: 'User', default: null },
    name: { type: String, default: '' },
  },
  diselesaikan_pada: { type: Date, default: null },

  nama_driver: { type: String, default: null },
  kontak_driver: { type: String, default: null },
  driver_user_id: { type: mongoose.Schema.Types.ObjectId, ref: 'User', default: null },

  tujuan_selanjutnya: {
    tipe: { type: String, enum: ['cabang', 'penerima', null], default: null },
    cabang_id: { type: mongoose.Schema.Types.ObjectId, ref: 'Cabang', default: null },
    nama: { type: String, default: null },
  },

  nama_penerima_akhir: { type: String, default: null },

  // Denormalized — lokasi cabang terakhir untuk query cepat tanpa $expr
  current_cabang_id: { type: mongoose.Schema.Types.ObjectId, ref: 'Cabang', default: null },

  no_manifest: { type: String, default: null },

  jenis_pembayaran: { type: String, enum: ['cash', 'cod', 'tempo'], default: 'cash' },
  status_pembayaran: { type: String, enum: ['unpaid', 'paid'], default: 'paid' },
  tempo_hari: { type: Number, default: 14 },
  pembayaran_dikonfirmasi_oleh: {
    user_id: { type: mongoose.Schema.Types.ObjectId, ref: 'User', default: null },
    name: { type: String, default: '' },
    role: { type: String, default: '' },
    cabang_id: { type: mongoose.Schema.Types.ObjectId, ref: 'Cabang', default: null },
  },
  pembayaran_dikonfirmasi_pada: { type: Date, default: null },
  cod_cabang_id: { type: mongoose.Schema.Types.ObjectId, ref: 'Cabang', default: null },

  tracking_logs: [trackingLogSchema],
}, { timestamps: true });

transactionSchema.index({ lokasi_penerima: '2dsphere' });

// 🔍 Performa query driver & admin
transactionSchema.index({ driver_user_id: 1 });
transactionSchema.index({ createdAt: -1 });
transactionSchema.index({ current_cabang_id: 1, status_saat_ini: 1 });
transactionSchema.index({ no_manifest: 1 });

// 🚀 Optimasi pencarian history untuk admin cabang
transactionSchema.index({ 'tracking_logs.lokasi.cabang_id': 1, 'tracking_logs.timestamp': -1 });

// 🚀 Optimasi pencarian history untuk driver
transactionSchema.index({ 'tracking_logs.driver_ditugaskan.user_id': 1, 'tracking_logs.timestamp': -1 });

// 🚀 Optimasi query $or di manifest detail (branch tracking_logs.no_manifest)
transactionSchema.index({ 'tracking_logs.no_manifest': 1, createdAt: 1 });

// 🚀 Optimasi recent contacts per cabang
transactionSchema.index({ 'created_by.cabang_id': 1, createdAt: -1 });

// 🚀 Optimasi aggregate analytics traffic per cabang
transactionSchema.index({ createdAt: -1, kode_gerai: 1 });

// 🚀 Optimasi aggregate analytics drivers — filter date range + elemMatch tracking_logs
// createdAt sebagai prefix agar $match date-range bisa pakai index sebelum $elemMatch
transactionSchema.index({ createdAt: -1, 'tracking_logs.status': 1, 'tracking_logs.driver_ditugaskan.user_id': 1 });

// 🚀 Optimasi pembayaran
transactionSchema.index({ jenis_pembayaran: 1, status_pembayaran: 1 });
transactionSchema.index({ cod_cabang_id: 1, status_pembayaran: 1 });
transactionSchema.index({ 'created_by.cabang_id': 1, jenis_pembayaran: 1, status_pembayaran: 1 });

// 🚀 Optimasi aggregate analytics berbasis rentang tanggal + jenis_masalah
// (traffic, per-cabang, routes-top, customers-top pakai $match createdAt + jenis_masalah)
transactionSchema.index({ createdAt: -1, jenis_masalah: 1 });

// 🚀 Optimasi /wajib-setor cashAgg: $match createdAt + jenis_pembayaran + created_by.cabang_id
transactionSchema.index({ createdAt: -1, jenis_pembayaran: 1, 'created_by.cabang_id': 1 });

// 🚀 Optimasi /wajib-setor asalAgg & lastMileAgg: $match pembayaran_dikonfirmasi_pada + jenis_pembayaran + status_pembayaran
transactionSchema.index({ pembayaran_dikonfirmasi_pada: -1, jenis_pembayaran: 1, status_pembayaran: 1 });

module.exports = mongoose.model('Transaction', transactionSchema);
